File: src/Sample/Plumix.Sample/Demos/General/ScrollbarDemoPage.cs
using Avalonia;
using Avalonia.Media;
using Plumix.Material;
using Plumix.Rendering;
using Plumix.Widgets;
using Plumix.UI;
using MaterialScrollbar = Plumix.Material.Scrollbar;

// Dart parity source (reference): dart_sample/lib/demos/general/scrollbar_demo_page.dart (exact sample parity)

namespace Plumix;

public sealed class ScrollbarDemoPage : StatefulWidget
{
    public override State CreateState()
    {
        return new ScrollbarDemoPageState();
    }
}

internal sealed class ScrollbarDemoPageState : State
{
    private ScrollController _materialController = null!;
    private ScrollController _rawController = null!;
    private bool _safeArea;
    private bool _rtl;
    private int _shapeIndex;

    private OutlinedBorder? ThumbShape => _shapeIndex switch
    {
        1 => new CircleBorder(side: new BorderSide(Colors.DarkBlue, 1)),
        2 => new BeveledRectangleBorder(borderRadius: BorderRadius.Circular(4)),
        3 => new RoundedRectangleBorder(borderRadius: BorderRadius.Only(
            topLeft: 8, bottomRight: 8)),
        _ => null,
    };

    public override void InitState()
    {
        _materialController = new ScrollController();
        _rawController = new ScrollController();
    }

    public override void Dispose()
    {
        _materialController.Dispose();
        _rawController.Dispose();

        base.Dispose();
    }

    public override Widget Build(BuildContext context)
    {
        return new Column(
            crossAxisAlignment: CrossAxisAlignment.Stretch,
            spacing: 10,
            children:
            [
                new Text("Scrollbar + RawScrollbar", fontSize: 20, color: Colors.Black),
                new Text(
                    "Material state theming/fade beside an always-visible raw track; both thumbs are draggable.",
                    fontSize: 14,
                    color: Colors.DimGray),
                new Row(
                    spacing: 4,
                    children:
                    [
                        new TextButton(
                            onPressed: () => SetState(() => _safeArea = !_safeArea),
                            child: new Text(_safeArea ? "Insets: on" : "Insets: off")),
                        new TextButton(
                            onPressed: () => SetState(() => _rtl = !_rtl),
                            child: new Text(_rtl ? "RTL" : "LTR")),
                        new TextButton(
                            onPressed: () => SetState(() => _shapeIndex = (_shapeIndex + 1) % 4),
                            child: new Text(new[] { "Ellipse", "Circle", "Beveled", "Corners" }[_shapeIndex])),
                    ]),
                new Expanded(
                    child: new Row(
                        spacing: 12,
                        children:
                        [
                            new Expanded(
                                child: BuildPane(
                                    "Material state theme",
                                    new ScrollbarTheme(
                                        data: new ScrollbarThemeData(
                                            trackVisibility: WidgetStateProperty<bool?>.ResolveWith(states =>
                                                states.Contains(WidgetState.Hovered)),
                                            thickness: WidgetStateProperty<double?>.ResolveWith(states =>
                                                states.Contains(WidgetState.Hovered) ? 12 : 8),
                                            thumbColor: WidgetStateProperty<Color?>.ResolveWith(states =>
                                                states.Contains(WidgetState.Dragged)
                                                    ? Color.Parse("#FF7B1FA2")
                                                    : Color.Parse("#FF1565C0"))),
                                        child: new MaterialScrollbar(
                                            controller: _materialController,
                                            child: BuildList(_materialController, "material"))))),
                            new Expanded(
                                child: BuildPane(
                                    "Raw + track",
                                    new Directionality(
                                        textDirection: _rtl ? TextDirection.Rtl : TextDirection.Ltr,
                                        child: new MediaQuery(
                                            data: MediaQuery.Of(context) with
                                            {
                                                Padding = _safeArea ? new Thickness(12, 24, 20, 40) : default,
                                            },
                                            child: new RawScrollbar(
                                                    controller: _rawController,
                                                    thumbVisibility: true,
                                                    trackVisibility: true,
                                                    thickness: 8,
                                                    radius: ThumbShape is null ? Radius.Elliptical(4, 8) : null,
                                                    shape: ThumbShape,
                                                    thumbColor: Color.Parse("#B3005E7A"),
                                                    trackColor: Color.Parse("#14005E7A"),
                                                    trackBorderColor: Color.Parse("#33005E7A"),
                                                    child: BuildList(_rawController, "raw")))))),
                        ])),
            ]);
    }

    private static Widget BuildPane(string label, Widget child) => new Column(
        crossAxisAlignment: CrossAxisAlignment.Stretch,
        spacing: 6,
        children:
        [
            new Text(label, fontSize: 13, color: Colors.DimGray),
            new Expanded(child: child),
        ]);

    private static Widget BuildList(ScrollController controller, string prefix) => ListView.Builder(
        itemCount: 70,
        controller: controller,
        itemExtent: 40,
        padding: new Thickness(10),
        itemBuilder: (_, index) => new Container(
            color: index % 2 == 0 ? Colors.White : Color.Parse("#FFF4F7FA"),
            padding: new Thickness(10, 8),
            child: new Text($"{prefix} row #{index}", fontSize: 13, color: Colors.Black)),
        addAutomaticKeepAlives: false);
}

File: dart_sample/lib/demos/general/scrollbar_demo_page.dart
import 'package:material_ui/material_ui.dart';

class ScrollbarDemoPage extends StatefulWidget {
  const ScrollbarDemoPage({super.key});

  @override
  State<ScrollbarDemoPage> createState() => _ScrollbarDemoPageState();
}

class _ScrollbarDemoPageState extends State<ScrollbarDemoPage> {
  late final ScrollController _materialController;
  late final ScrollController _rawController;
  bool _safeArea = false;
  bool _rtl = false;
  int _shapeIndex = 0;

  OutlinedBorder? get _thumbShape => switch (_shapeIndex) {
    1 => const CircleBorder(side: BorderSide(color: Color(0xFF00008B))),
    2 => const BeveledRectangleBorder(
      borderRadius: BorderRadius.all(Radius.circular(4)),
    ),
    3 => const RoundedRectangleBorder(
      borderRadius: BorderRadius.only(
        topLeft: Radius.circular(8),
        bottomRight: Radius.circular(8),
      ),
    ),
    _ => null,
  };

  @override
  void initState() {
    super.initState();
    _materialController = ScrollController();
    _rawController = ScrollController();
  }

  @override
  void dispose() {
    _materialController.dispose();
    _rawController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      spacing: 10,
      children: <Widget>[
        const Text(
          'Scrollbar + RawScrollbar',
          style: TextStyle(fontSize: 20, color: Colors.black),
        ),
        const Text(
          'Material state theming/fade beside an always-visible raw track; both thumbs are draggable.',
          style: TextStyle(fontSize: 14, color: Colors.black54),
        ),
        Row(
          spacing: 4,
          children: <Widget>[
            TextButton(
              onPressed: () => setState(() => _safeArea = !_safeArea),
              child: Text(_safeArea ? 'Insets: on' : 'Insets: off'),
            ),
            TextButton(
              onPressed: () => setState(() => _rtl = !_rtl),
              child: Text(_rtl ? 'RTL' : 'LTR'),
            ),
            TextButton(
              onPressed: () =>
                  setState(() => _shapeIndex = (_shapeIndex + 1) % 4),
              child: Text(
                ['Ellipse', 'Circle', 'Beveled', 'Corners'][_shapeIndex],
              ),
            ),
          ],
        ),
        Expanded(
          child: Row(
            spacing: 12,
            children: <Widget>[
              Expanded(
                child: _buildPane(
                  'Material state theme',
                  ScrollbarTheme(
                    data: ScrollbarThemeData(
                      trackVisibility: WidgetStateProperty.resolveWith<bool?>(
                        (Set<WidgetState> states) =>
                            states.contains(WidgetState.hovered),
                      ),
                      thickness: WidgetStateProperty.resolveWith<double?>(
                        (Set<WidgetState> states) =>
                            states.contains(WidgetState.hovered) ? 12 : 8,
                      ),
                      thumbColor: WidgetStateProperty.resolveWith<Color?>(
                        (Set<WidgetState> states) =>
                            states.contains(WidgetState.dragged)
                            ? const Color(0xFF7B1FA2)
                            : const Color(0xFF1565C0),
                      ),
                    ),
                    child: Scrollbar(
                      controller: _materialController,
                      child: _buildList(_materialController, 'material'),
                    ),
                  ),
                ),
              ),
              Expanded(
                child: _buildPane(
                  'Raw + track',
                  Directionality(
                    textDirection: _rtl ? TextDirection.rtl : TextDirection.ltr,
                    child: MediaQuery(
                      data: MediaQuery.of(context).copyWith(
                        padding: _safeArea
                            ? const EdgeInsets.fromLTRB(12, 24, 20, 40)
                            : EdgeInsets.zero,
                      ),
                      child: RawScrollbar(
                        controller: _rawController,
                        thumbVisibility: true,
                        trackVisibility: true,
                        thickness: 8,
                        radius: _thumbShape == null
                            ? const Radius.elliptical(4, 8)
                            : null,
                        shape: _thumbShape,
                        thumbColor: const Color(0xB3005E7A),
                        trackColor: const Color(0x14005E7A),
                        trackBorderColor: const Color(0x33005E7A),
                        child: _buildList(_rawController, 'raw'),
                      ),
                    ),
                  ),
                ),
              ),
            ],
          ),
        ),
      ],
    );
  }

  Widget _buildPane(String label, Widget child) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      spacing: 6,
      children: <Widget>[
        Text(
          label,
          style: const TextStyle(fontSize: 13, color: Colors.black54),
        ),
        Expanded(child: child),
      ],
    );
  }

  Widget _buildList(ScrollController controller, String prefix) {
    return ListView.builder(
      controller: controller,
      itemCount: 70,
      itemExtent: 40,
      padding: const EdgeInsets.all(10),
      itemBuilder: (BuildContext context, int index) {
        return Container(
          color: index.isEven ? Colors.white : const Color(0xFFF4F7FA),
          padding: const EdgeInsets.symmetric(horizontal: 10, vertical: 8),
          child: Text(
            '$prefix row #$index',
            style: const TextStyle(fontSize: 13, color: Colors.black),
          ),
        );
      },
    );
  }
}
